### PR TITLE
Push Notifications

### PR DESCRIPTION
**Pull request by mazenchami**
# Description

[Trello Card](10-push-notifications)

This PR hooks up our PNs via Firebase Messaging console.

# Graphics

| iOS | Android |
| ------ | ------- |
|  | android_screenshot |

# Checklist:

- [x] I have done a thorough self-review of my code
- [ ] I have tested with a screen reader and font-scaling turned on and added necessary accessibility features 
- [x] I have run tests and linter

### CODE REVIEW

File: App.js [hunk 1]
```diff
@@ -4,6 +4,12 @@ import App from "./app/app.tsx"
 import React from "react"
 import { registerRootComponent } from "expo"
 import * as SplashScreen from "expo-splash-screen"
+import messaging from "@react-native-firebase/messaging"
+
+// Register background handler
+messaging().setBackgroundMessageHandler(async remoteMessage => {
+  console.log('Message handled in the background!', remoteMessage);
```
Comment (frankcalise) [inline App.js:11]: ✂️

File: app.config.ts [hunk 2]
```diff
@@ -1,71 +1,75 @@
...
-      tabletImage: './assets/images/splash-logo-ios-tablet.png',
-      resizeMode: 'contain',
-      backgroundColor: '#081828'
+      image: "./assets/images/splash-logo-ios-mobile.png",
+      tabletImage: "./assets/images/splash-logo-ios-tablet.png",
+      resizeMode: "contain",
+      backgroundColor: "#081828",
+    },
+    googleServicesFile: `./GoogleService-Info.plist`,
+    infoPlist: {
```
Comment (frankcalise) [inline app.config.ts:51]: Probably going to have a merge conflict here due to the hidden status bar but should be an easy merge.

File: app/screens/WelcomeScreen.tsx [hunk 3]
```diff
@@ -1,23 +1,38 @@
...
 const { width: screenWidth } = Dimensions.get("screen")
 
 interface WelcomeScreenProps extends AppStackScreenProps<"Welcome"> {}
 
+const requestUserPermission = async () => {
+  const authStatus = await messaging().requestPermission()
+  const enabled =
+    authStatus === messaging.AuthorizationStatus.AUTHORIZED ||
+    authStatus === messaging.AuthorizationStatus.PROVISIONAL
+  if (enabled) {
```
Comment (frankcalise) [inline app/screens/WelcomeScreen.tsx:20]: Should we alert anything if they have denied permissions and come back to this screen?

**Review by frankcalise — COMMENTED** [review]
Added some notes

File: app/screens/WelcomeScreen.tsx (hunk 3 above)
Comment (mazenchami) [reply app/screens/WelcomeScreen.tsx:20]: Honestly, in my past experience usually if a user doesn't accept one of these modals, its a burden to keep coming back to it. We might just leave it as is but also have a convo with the rest of the team if we want to go down a specific route

File: app/screens/InfoScreen.tsx [hunk 4]
```diff
@@ -33,6 +42,17 @@ export const InfoScreen: React.FunctionComponent<TabScreenProps<"Info">> = () =>
...
+      <Button
+        onPress={async () => {
+          const token = await messaging().getToken()
+          Alert.alert("PN Token", token)
+        }}
+        preset="link"
+        style={$hiddenButton}
+        TextProps={{ maxFontSizeMultiplier: 1.8 }}
+      />
+    ),
```
[the remark this answers is not in the document]
Comment (frankcalise) [reply app/screens/InfoScreen.tsx:55]: hidden as public as this repo is :D

File: app/app.tsx [hunk 5]
```diff
@@ -46,6 +54,38 @@ interface AppProps {
...
+// Setting up our custom Toast component
+const CustomToast = () => {
+  const insets = useSafeAreaInsets()
+
+  useLayoutEffect(() => {
+    // handle a new push notification received while the app is in "foreground" state
+    const unsubscribe = messaging().onMessage(async (remoteMessage) => {
+      if (remoteMessage.notification) {
+        Toast.show({
+          text1: remoteMessage.notification.title,
```
Comment (frankcalise) [inline app/app.tsx:66]: Does the Firebase dashboard require title and body? Or at least one? 

If not, maybe here we should check if both are empty so we don't send out blank push notifications

**Review by frankcalise — APPROVED** [review]
Looking good! Just one question on some of the title/body fields and Firebase

File: app/app.tsx (hunk 5 above)
Comment (mazenchami) [reply app/app.tsx:69]: good point. updated the logic here: [955300f](https://github.com/infinitered/ChainReactApp2023/pull/36/commits/955300f3b195ed60d5449a26734e9af839b69236)

`Toast` does a good job of handling if one or the other is `undefined`.

**Review by robinheinze — APPROVED** [review]
Looks great!

**Pull request merged**